### ISSUE TITLE
[RF] RooFormulaVar not working in Root 6.26

### ISSUE THREAD

**Issue by EmilRosanowski** [root-project/root#11760]
### Explain what you would like to see improved

I tried to shift a RooHistPdf by using RooFormulaVar. This lead to the following problems:
RooFit cannot normalize the RooHistPdf correctly (always some random heights)
RooFit cannot fit the RooHistPdf to a RooArgSet or RooDataHist and a lot of evaluation errors occur.

As soon as I let my code run on root 6.24 there is no problem anymore and both things work. Therefor I suspect a problem with Root 6.26

### Optional: share how it could be improved

The combination of RooFormulaVar and RooHistPdf seems to bring up problems.

### To Reproduce

```C++
using namespace RooFit;

void test2()
{
    RooRealVar x("x","x",800,1800);
    RooPlot *frame = x.frame();

    RooRealVar mean("mean","mean",1200,1000,1400);
    RooRealVar sigma("sigma","sigma",80,50,90);
    RooGaussian Gauss("Gauss","Gauss",x,mean,sigma);


    RooRealVar mean3("mean3","mean3",1400,1200,1600);
    RooRealVar sigma3("sigma3","sigma3",80,50,90);
    RooGaussian Gauss3("Gauss3","Gauss3",x,mean3,sigma3);

    RooRealVar part1("p1","p1",1000.,0.,1000.);
    RooRealVar part2("p2","p2",1000,0,1000);

    RooAddPdf Data_Model("DM","DM",RooArgList(Gauss,Gauss3),RooArgList(part1,part2));

    RooDataSet *Data = Data_Model.generate(x,10000);

    Data_Model.plotOn(frame, LineColor(kBlack));
    Data->plotOn(frame, MarkerColor(kRed));

    RooRealVar mean2("mean2","mean",1210,1200,1220);
    RooRealVar sigma2("sigma2","sigma",80,50,90);
    RooGaussian Gauss2("Gauss2","Gauss",x,mean2,sigma2);

    RooRealVar mean4("mean4","mean4",1410,1200,1600);
    RooRealVar sigma4("sigma4","sigma4",80,50,90);
    RooGaussian Gauss4("Gauss4","Gauss4",x,mean4,sigma4);

    RooDataSet *Model1 = Gauss2.generate(x,1000000);

    RooDataSet *Model2 = Gauss4.generate(x,1000000);

    //Model->plotOn(frame, LineColor(kGreen), MarkerColor(kGreen));

    RooRealVar Shift("Shift","Shift",0,-10,10);
    RooFormulaVar xShifted("xShifted","x + Shift",RooArgSet(x,Shift));

    RooRealVar Shift2("Shift2","Shift2",0,-10,10);
    RooFormulaVar xShifted2("xShifted2","x + Shift2",RooArgSet(x,Shift2));

    RooDataHist Sim("Sim","Sim",x,*Model1);
    RooDataHist Sim2("Sim2","Sim2",x,*Model2);

    RooHistPdf Func("Func","func",xShifted,x,Sim);
    RooHistPdf Func2("Func2","func2",xShifted2,x,Sim2);

    RooRealVar Amount("A","A",10000,0,2000000);
    RooRealVar Amount2("A2","A2",10000,0,2000000);

    RooAddPdf Func1("f","f",RooArgList(Func,Func2),RooArgList(Amount,Amount2));

    Func1.plotOn(frame, LineColor(kRed));
    std::cout<<Func.getNorm()<<std::endl;
    Func1.fitTo(*Data, Range(1000,1600));
    Func1.plotOn(frame);
    std::cout<<Func.getNorm()<<std::endl;
    //Normalization(2500000,RooAbsReal::NumEvent)
    Func1.paramOn(frame);
    frame->Draw();
}
```

Sorry, its ugly code, but I am not really good at programming.

### Setup

Root Version 6.26/06 and 6.24/06
Operating System: MacOS Monterey and Linux respectively
Obtained via homebrew

### Additional context

homebrew does not yet include root 6.28 and thus I cannot install it to check if the mistake appeared there to.

**Comment by jalopezg-r00t** [comment]
For context, see https://root-forum.cern.ch/t/normalization-of-pdfs-with-rooformulavar/52485.

**Comment by guitargeek** [comment]
Hi! Thanks for opening the issue. I have first added the `priority-high` label because it appeared like a usecase that worked in 6.24 was broken in 6.26, but I don't this is the case. Yes, the normalization was not off in 6.24, but the fit gave nonsensical results anyway for different reasons also in 6.24. Or did you manage to make this shifted RooHistPdf work in any ROOT version?

To me, the result of your script looks like this both in ROOT 6.24 and in ROOT `master` (upcoming 6.28, where the normalization issue got fixed by the way):

![plot](https://user-images.githubusercontent.com/6578603/203787998-2652e495-f6b6-4d54-a2fc-64c66a96ac8a.png)

I modified your script a bit to have more events in the toy dataset, and fixed the plotting order (you got to plot the data always the very first, otherwise the PDFs will not be scaled correctly).

I think the best we can do now to support your usecase is this:

1. Since shifting a RooHistPdf doesn't work in any ROOT version, I implement a custom PDF class that can do that which you can readily use
2. For the upcoming ROOT 6.28, we will also make sure that the actual RooHistPdf support the shift
3. About this issue with the normalization in 6.26, I'm not sure if it's worth it to address this since the problem is gone in ROOT 6.28 which will be released soon. Anyway, the problem only appears for models where the fit result was also wrong before for other reasons, so that temporary regression in 6.26 probably didn't break any existing usecase.

Does that make sense to you?

**Comment by EmilRosanowski** [comment]
Hi, thanks for your support. In Root 6.24 everything works well and looks exactly like the plot you attached. However if the problem is not there anymore in root 6.28 it is probably not worth to do anything about this as this is a very specific and small problem. As I also have a linux computer with root 6.24 it should not be too big a problem for me. But again, thank you very much for taking a look at my problem.
Thanks,
Emil

**Comment by guitargeek** [comment]
Okay! But actually it doesn't work also in 6.24 no? The plot that I attached shows that the fit does the wrong thing: instead of shifting the red template to match the data generated according to the black curve, the component normalizations are fitted such that the tails are modeled fine, but in the middle the fit represented by the blue line is completely off because no shift on the x-axis happened.

So we still need to fix the problem with the shifted RooHistPdf before you can do your work, or do I understand something wrongly?

**Comment by guitargeek** [comment]
Okay great, thank you very much for clarifying your usecase!

I'll close this issue then after I implemented your code as a unit test in ROOT such that the normalization is not allowed to break again.

**Comment by guitargeek** [comment]
I finally figured out what the problem was in 6.26 specifically. In the numeric integration of the PDF, the new BatchMode was used, but it still had some bugs in ROOT 6.26. That's why we didn't see the problem in 6.24. And in ROOT master, we also didn't see it because the problem with the BatchMode got fixed.

To fix the problem also in 6.26, I opened a PR where the BatchMode is disabled for the numeric integration of the PDF:
https://github.com/root-project/root/pull/11961
